### ISSUE TITLE
Writing XML async to HttpResponseStream

### ISSUE THREAD

**Issue by Tornhoof** [aspnet/AspNetCore#14811]
### Describe the bug
The check for AllowSynchronousIO is too strict and  throws even if no data is written.

The problem is related to the ``XmlWellFormedWriter`` from Corefx flushing the data on close() in all cases, even if you flush the data yourself.
https://github.com/dotnet/corefx/blob/c1092870c306b8da256e455cec0814486255d0c1/src/System.Private.Xml/src/System/Xml/Core/XmlWellFormedWriter.cs#L1266
If you take a look at the flush implementation in ``XmlUtf8RawTextWriter``.
https://github.com/dotnet/corefx/blob/c1092870c306b8da256e455cec0814486255d0c1/src/System.Private.Xml/src/System/Xml/Core/XmlUtf8RawTextWriter.cs#L674
You see that it writes to stream regardless of the bufPos, a few lines below the bufPos is set to 1 https://github.com/dotnet/corefx/blob/c1092870c306b8da256e455cec0814486255d0c1/src/System.Private.Xml/src/System/Xml/Core/XmlUtf8RawTextWriter.cs#L701
That means even if the bufPos is 1, the write method is called with length 0.

Unfortunately this poses a problem in ```HttpResponseStream`` (Kestrel/IIS):
https://github.com/aspnet/AspNetCore/blob/354d859d50f252526be5a6264434d1765df86450/src/Servers/Kestrel/Core/src/Internal/Http/HttpResponseStream.cs#L80-L83
There the SyncIO Check is done without checking for the length argument, meaning it always throws.
This basically means, that at the moment it's not possible to write Xml to the HttpResponseStream without enabling SyncIo. This is quite annoying at the moment.

Possible solutions:
1. Make the things proper Async Dispose (complex)
2. Change XmlUtf8RawTextWriter to only write to the stream if length > 0 (easy, in corefx), but usually that check is usually an implementation detail
3. Change those SyncIO Checks to only throw on length > 0 (easy, that's why I opened the bug here)

The third way is currently what I'm trying in my solution, instead of using bodystream directly, I use a wrapperstream checking for the length == 0 case.

### To Reproduce
1. Using this version of ASP.NET Core '3.0'
2. Run this code (see repro below)
3. With these arguments: none

### Expected behavior
No exception, it works.
### Additional context

Repro:
```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Xml;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace XmlWriterRepro
{
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder => { webBuilder.UseStartup<Startup>(); });
    }

    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();

            app.UseEndpoints(endpoints => { endpoints.MapGet("/", async context => { await WriteXmlAsync(context.Response.Body, "Hello World"); }); });
        }

        private static async Task WriteXmlAsync(Stream output, string text)
        {
            using (var xmlWriter = XmlWriter.Create(output, new XmlWriterSettings {Async = true, CloseOutput = false, WriteEndDocumentOnClose = false, CheckCharacters = false}))
            {
                await xmlWriter.WriteStartDocumentAsync().ConfigureAwait(false);
                await xmlWriter.WriteElementStringAsync(null, "Value", null, text).ConfigureAwait(false);
                await xmlWriter.WriteEndDocumentAsync().ConfigureAwait(false);
                await xmlWriter.FlushAsync().ConfigureAwait(false);
            }
        }
    }
}

```

**Comment by davidfowl** [comment]
> Change those SyncIO Checks to only throw on count > 0 (easy, that's why I opened the bug here)

I'm not sure we'd make this change, it's a bit edgy and hard to explain. If you need to do synchronous IO, then you should buffer the response output and copy asynchronously.

**Comment by Tornhoof** [comment]
Yeah, I understand that, but the code does not need sync, It's an implementation issue in the xmlwriters which simply are not implementing async properly (doing "work" in Close()). The Wrapper Stream I use solves that problem nicely enough, because after writing the enddocument tag and flushing the data there is nothing to do there.

If you prefer, I can close this and reopen a bug in corefx for tracking the "work" in Close().

**Comment by Tornhoof** [comment]
Closing this.
